### ISSUE TITLE
[BUG] multi-column groupby with "count" agg fails

### ISSUE THREAD

**Issue by randerzander** [rapidsai/dask-cudf#231]
Now that multiindex support merged into cuDF, I was testing some multi-column groupbys in dask-cudf and ran into the following issue.

From the tip of cudf and dask-cudf branch-0.7:
```
import cudf, dask_cudf

tmp_df = cudf.DataFrame()
tmp_df['id'] = [0, 0, 1, 2, 2, 2]
tmp_df['val1'] = [0, 1, 0, 0, 1, 2]
tmp_df['val2'] = [9, 9, 9, 9, 9, 9]

ddf = dask_cudf.from_cudf(tmp_df, npartitions=1)
ddf.groupby(['id', 'val1']).count().compute()
```
Result:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-34-24268c614215> in <module>
      7 
      8 ddf = dask_cudf.from_cudf(tmp_df, npartitions=1)
----> 9 ddf.groupby(['id', 'val1']).count().compute()

/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/groupby.py in count(self, split_every, split_out)
    961         return self._aca_agg(token='count', func=M.count,
    962                              aggfunc=M.sum, split_every=split_every,
--> 963                              split_out=split_out)
    964 
    965     @derived_from(pd.core.groupby.GroupBy)

/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/groupby.py in _aca_agg(self, token, func, aggfunc, split_every, split_out)
    835                    meta=meta, token=token, split_every=split_every,
    836                    aggregate_kwargs=dict(aggfunc=aggfunc, levels=levels),
--> 837                    split_out=split_out, split_out_setup=split_out_on_index)
    838 
    839     def _cum_agg(self, token, chunk, aggregate, initial):

/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/core.py in apply_concat_apply(args, chunk, aggregate, combine, meta, token, chunk_kwargs, aggregate_kwargs, combine_kwargs, split_every, split_out, split_out_setup, split_out_setup_kwargs, **kwargs)
   3723                         **aggregate_kwargs)
   3724     meta = make_meta(meta, index=(getattr(make_meta(dfs[0]), 'index', None)
-> 3725                                   if dfs else None))
   3726 
   3727     graph = HighLevelGraph.from_collections(b, dsk, dependencies=dfs)

/conda/envs/rapids/lib/python3.7/site-packages/dask/utils.py in __call__(self, arg, *args, **kwargs)
    415         """
    416         meth = self.dispatch(type(arg))
--> 417         return meth(arg, *args, **kwargs)
    418 
    419     @property

/conda/envs/rapids/lib/python3.7/site-packages/dask_cudf-0.0.0.dev0-py3.7.egg/dask_cudf/backends.py in make_meta_cudf(x, index)
     19 @make_meta.register((cudf.Series, cudf.DataFrame))
     20 def make_meta_cudf(x, index=None):
---> 21     return x.head(0)
     22 
     23 

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0.dev0+1448.gb936b527.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/dataframe.py in head(self, n)
    358         1    1  11.0
    359         """
--> 360         return self.iloc[:n]
    361 
    362     def tail(self, n=5):

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0.dev0+1448.gb936b527.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/dataframe.py in __getitem__(self, arg)
   2891             df = DataFrame()
   2892             for col in self._df.columns:
-> 2893                 df[col] = self._df[col][arg]
   2894             df.index = self._df.index[arg]
   2895 

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0.dev0+1448.gb936b527.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/series.py in __getitem__(self, arg)
    284 
    285         elif isinstance(arg, slice):
--> 286             index = self.index[arg]         # slice index
    287             col = self._column[arg]         # slice column
    288             return self._copy_construct(data=col, index=index)

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0.dev0+1448.gb936b527.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/multiindex.py in __getitem__(self, index)
    237 
    238     def __getitem__(self, index):
--> 239         match = self.take(index)
    240         result = []
    241         for level, item in enumerate(match.codes):

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0.dev0+1448.gb936b527.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/multiindex.py in take(self, indices)
    219         elif isinstance(indices, Series):
    220             indices = indices.to_gpu_array()
--> 221         codes = self.codes.take(indices)
    222         result = MultiIndex(self.levels, codes)
    223         result.names = self.names

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0.dev0+1448.gb936b527.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/dataframe.py in take(self, positions, ignore_index)
    788         out = DataFrame()
    789         for col in self.columns:
--> 790             out[col] = self[col].take(positions, ignore_index=ignore_index)
    791         return out
    792 

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0.dev0+1448.gb936b527.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/series.py in take(self, indices, ignore_index)
    300             indices = indices.to_gpu_array()
    301         else:
--> 302             indices = Buffer(indices).to_gpu_array()
    303         # Handle zero size
    304         if indices.size == 0:

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.0.dev0+1448.gb936b527.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/buffer.py in __init__(self, mem, size, capacity, categorical)
     30                 size = len(mem)
     31             else:
---> 32                 size = mem.size
     33         if capacity is None:
     34             capacity = size

AttributeError: 'slice' object has no attribute 'size'
```

**Comment by thomcom** [comment]
I working on making this issue go away. :)
Unfortunately the exposed behavior of groupby between cudf and pandas is identical and works as expected:

```
import cudf, dask_cudf
import pandas as pd

pdf = pd.DataFrame()
pdf['id'] = [0, 0, 1, 2, 2, 2]
pdf['val1'] = [0, 1, 0, 0, 1, 2]
pdf['val2'] = [9, 9, 9, 9, 9, 9]

gdf = cudf.from_pandas(pdf)

print(pdf.groupby(['id', 'val1']).count())
print(gdf.groupby(['id', 'val1']).count().to_pandas())
```
```output
         val2
id val1
0  0        1
   1        1
1  0        1
2  0        1
   1        1
   2        1
         val2
id val1
0  0        1
   1        1
1  0        1
2  0        1
   1        1
   2        1
```

I'm working on understanding the metadata groupby performed by dask so I can better support it in cudf.

**Comment by mrocklin** [comment]
I've also been playing with this, but so far I'm mostly stuck on getting a good environment up with cudf nightly builds.  Hopefully I'll also be able to guide here shortly?

**Comment by thomcom** [comment]
Ok, so I know what I'm working on today. :) There's a whole set of semi-hidden behaviors pandas uses for indexes when, you guessed it, the groupby result is empty. Turns out you depend extensively on them for dask-cudf. :D I should be able to support the set of empty groupby/index results that pandas produces in cudf by end of day. Probably before - tests are written, I'm working through them now.

**Comment by mrocklin** [comment]
I take it that the current code paths don't work with empty dataframes?

If you're making different code paths for empty dataframes then I would become slightly concerned about a maintenance issue (every time we make a change we now need to make another change).  Is there some way to generalize the original code path?

**Comment by thomcom** [comment]
I don't think it is an inappropriate modification. `libcudf` performs groupbys without any knowledge of indices - all of the data columns are present in the input and output to libcudf groupby. After the groupby is complete, I wrote a post-processing step that, using the structure and shape of the _by arguments to construct an output dataframe and index that match Pandas. I basically ignored the dataframe result returning no rows case. Now I'm having to write that case. I could modify the existing code to handle empty cases more gracefully, or add a condition for empty cases and handle them specifically. I'm not sure which will be easier to maintain in the long run, but I know which is easier to write today.

**Comment by thomcom** [comment]
I think :)

**Comment by thomcom** [comment]
So I corrected the code to handle empty groupbys, but a major issue remains with trying to do `.agg('count')` that seems to be unrelated:

```
>>> ddf.groupby(['id', 'val1']).agg('sum')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 1194, in agg
    return self.aggregate(arg, split_every=split_every, split_out=split_out)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 1190, in aggregate
    return super(DataFrameGroupBy, self).aggregate(arg, split_every=split_every, split_out=split_out)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 1067, in aggregate
    split_out=split_out, split_out_setup=split_out_on_index)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/core.py", line 3656, in apply_concat_apply
    meta_chunk = _emulate(chunk, *args, udf=True, **chunk_kwargs)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/core.py", line 3697, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/utils.py", line 177, in raise_on_meta_error
    raise ValueError(msg)
ValueError: Metadata inference failed in `_groupby_apply_funcs`.

You have supplied a custom function and Dask is unable to
determine the type of output that that function returns.

To resolve this please provide a meta= keyword.
The docstring of the Dask function you ran should have more information.

Original error is below:
------------------------
ValueError('If using all scalar values, you must pass an index')

Traceback:
---------
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/utils.py", line 160, in raise_on_meta_error
    yield
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/core.py", line 3697, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 663, in _groupby_apply_funcs
    return pd.DataFrame(result)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/pandas/core/frame.py", line 348, in __init__
    mgr = self._init_dict(data, index, columns, dtype=dtype)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/pandas/core/frame.py", line 459, in _init_dict
    return _arrays_to_mgr(arrays, data_names, index, columns, dtype=dtype)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/pandas/core/frame.py", line 7356, in _arrays_to_mgr
    index = extract_index(arrays)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/pandas/core/frame.py", line 7393, in extract_index
    raise ValueError('If using all scalar values, you must pass'

Metadata inference failed in `_groupby_apply_funcs`.

You have supplied a custom function and Dask is unable to
determine the type of output that that function returns.

To resolve this please provide a meta= keyword.
The docstring of the Dask function you ran should have more information.

Original error is below:
------------------------
ValueError('If using all scalar values, you must pass an index')

Traceback:
---------
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/utils.py", line 160, in raise_on_meta_error
    yield
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/core.py", line 3697, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 663, in _groupby_apply_funcs
    return pd.DataFrame(result)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/pandas/core/frame.py", line 348, in __init__
    mgr = self._init_dict(data, index, columns, dtype=dtype)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/pandas/core/frame.py", line 459, in _init_dict
    return _arrays_to_mgr(arrays, data_names, index, columns, dtype=dtype)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/pandas/core/frame.py", line 7356, in _arrays_to_mgr
    index = extract_index(arrays)
  File "/home/nfs/tcomer/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/pandas/core/frame.py", line 7393, in extract_index
    raise ValueError('If using all scalar values, you must pass'
```

**Comment by mrocklin** [comment]
Using `pdb` to look at the state where the exception is happening, it appears to be doing something like the following:

```python
ipdb> pp cudf.Series([1, 2, 3]).take(slice(None))
*** AttributeError: 'slice' object has no attribute 'size'
```

**Comment by mrocklin** [comment]
On the following code example

```python
import cudf, dask_cudf

tmp_df = cudf.DataFrame()
tmp_df['id'] = [0, 0, 1, 2, 2, 2]
tmp_df['val1'] = [0, 1, 0, 0, 1, 2]
tmp_df['val2'] = [9, 9, 9, 9, 9, 9]

ddf = dask_cudf.from_cudf(tmp_df, npartitions=1)
ddf.groupby(['id', 'val1']).agg('sum')
```

The traceback shows that Dask is calling `pd.DataFrame` on something

```python-traceback
~/dask/dask/dataframe/core.py in _emulate(func, *args, **kwargs)
   3763     """
   3764     kwargs.pop('udf')
-> 3765     return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
   3766
   3767

~/dask/dask/dataframe/groupby.py in _groupby_apply_funcs(df, *index, **kwargs)
    672             result[result_column] = r
    673
--> 674     return pd.DataFrame(result)
    675
    676

~/miniconda/envs/cudf-nightly/lib/python3.7/site-packages/pandas/core/frame.py in __init__(self, data, index, columns, dtype, copy)
    390                                  dtype=dtype, copy=copy)
    391         elif isinstance(data, dict):
--> 392             mgr = init_dict(data, index, columns, dtype=dtype)
    393         elif isinstance(data, ma.MaskedArray):
    394             import numpy.ma.mrecords as mrecords
```

If I make this change

```diff
diff --git a/dask/dataframe/groupby.py b/dask/dataframe/groupby.py
index 6fcadce..c8e101b 100644
--- a/dask/dataframe/groupby.py
+++ b/dask/dataframe/groupby.py
@@ -671,7 +671,7 @@ def _groupby_apply_funcs(df, *index, **kwargs):
         else:
             result[result_column] = r

-    return pd.DataFrame(result)
+    return type(df)(result)


 def _compute_sum_of_squares(grouped, column):
```

Then we get here:

```python
> /home/nfs/mrocklin/cudf/python/cudf/dataframe/dataframe.py(892)_prepare_series_for_add()
    890
    891         empty_index = len(self._index) == 0
--> 892         if forceindex or empty_index or self._index.equals(series.index):
    893             if empty_index:
    894                 self._index = series.index

ipdb> pp forceindex
False
ipdb> pp empty_index
False
ipdb> pp index
*** NameError: name 'index' is not defined
ipdb> pp series.index
GenericIndex([0], dtype=int64)
ipdb> pp self._index
MultiIndex(levels=[<cudf.Series nrows=1 >, <cudf.Series nrows=1 >],
codes=   id  val1
0   0     0)
```

So perhaps that check within cudf could use some review

**Comment by mrocklin** [comment]
See https://github.com/dask/dask/pull/4778 to get us past the dask issues.  The index comparison issues seem real though?

**Comment by thomcom** [comment]
That check has ruined my day repeatedly.

**Comment by mrocklin** [comment]
Yeah, @quasiben and I ran into it too when playing with cuML.  It can be quite expensive in the common case.

**Comment by quasiben** [comment]
@randerzander I believe we can close this now with https://github.com/rapidsai/cudf/pull/1659 merged in.  Can you confirm this works for you, and if so, close ?

```
In [1]: import cudf, dask_cudf
   ...:
   ...: tmp_df = cudf.DataFrame()
   ...: tmp_df['id'] = [0, 0, 1, 2, 2, 2]
   ...: tmp_df['val1'] = [0, 1, 0, 0, 1, 2]
   ...: tmp_df['val2'] = [9, 9, 9, 9, 9, 9]
   ...:
   ...: ddf = dask_cudf.from_cudf(tmp_df, npartitions=1)
   ...: print(ddf.groupby(['id', 'val1']).count().compute().to_pandas())
   ...: print(tmp_df.groupby(['id', 'val1']).count().to_pandas())
         val2
id val1
0  0        1
   1        1
1  0        1
2  0        1
   1        1
   2        1
         val2
id val1
0  0        1
   1        1
1  0        1
2  0        1
   1        1
   2        1

In [2]: from cudf.tests.utils import assert_eq

In [3]: import dask.dataframe as dd

In [4]: dd.assert_eq(tmp_df.groupby(['id', 'val1']).count(), ddf.groupby(['id', 'val1']).count())
Out[4]: True
```

**Comment by quasiben** [comment]
closing. @randerzander if you are still getting the error please reopen

**Comment by randerzander** [comment]
The issue appears to have returned in the latest 0.7 nightlies:
```
dev@test-m:~$ conda list | grep cudf
cudf                      0.7.0.dev0            py37_1568    rapidsai-nightly/label/cuda10.0
dask-cudf                 0.7.0.dev0              py37_37    rapidsai-nightly/label/cuda10.0
libcudf                   0.7.0.dev0        cuda10.0_1568    rapidsai-nightly/label/cuda10.0
```

**Comment by thomcom** [comment]
Thanks Randy. This is a different issue than the one originally filed under this bug. The issue here is related to dask-cudf and that a dask-cudf groupby in this case is producing a `SeriesGroupBy` object from the multiple groupby, when CUDF and Pandas both produce a DataFrameGroupby (CUDF just calls this a Groupby). Pandas SeriesGroupBy and DataFrameGroupby are basically identical, so dask-cudf never noticed, but unfortunately CUDF SeriesGroupBy isn't as fully featured as the dataframe Groupby. If @quasiben can make the internals of dask use the Groupby object instead of the SeriesGroupBy this should work again.

**Comment by quasiben** [comment]
I believe this should be resolved by https://github.com/dask/dask/pull/4786

**Comment by randerzander** [comment]
This is resolved with cudf branch `release-0.7.2`.

Thanks everyone for the hard work figuring this out!
